StringBuilder.AppendJoin (appending lists to StringBuilder)

#### Background

Let's say you want to build a string that contains a separated list (e.g. list of filenames, messages, or anything else).

If it is the only thing you want, you can use `string.Join`. However what if you want to append something else? If you `Concat` with the result of `string.Join`, that means that you unnecessarily allocated an intermediate joined string, which can be rather large (depending on the original collection).

On the other hand, `StringBuilder` would require something like:

``` csharp
var builder = new StringBuilder("Names: ");
for (var i = 0; i < names.Length; i++) {
    if (i > 0) { // this becomes even more involved if `names` is `IEnumerable`
        builder.Append(", ");
    }
    builder.Append(names[i]);
}
```

That can be abstracted into an extension method, however I believe that extension would be general and useful enough to benefit from being in NetFX.
#### API Design

(updated based on API review, second pass -- see https://github.com/dotnet/corefx/issues/3419#issuecomment-273254581)

``` csharp
public class StringBuilder {
    public StringBuilder AppendJoin(string separator, params object[] values);
    public StringBuilder AppendJoin<T>(string separator, IEnumerable<T> values);
    public StringBuilder AppendJoin(string separator, params string[] value);

    public StringBuilder AppendJoin(chat separator, params object[] values);
    public StringBuilder AppendJoin<T>(char separator, IEnumerable<T> values);
    public StringBuilder AppendJoin(char separator, params string[] value);

    // string.Join overloads NOT taken:
    //public StringBuilder AppendJoin(string separator, IEnumerable<string> values);
    //public StringBuilder AppendJoin(string separator, string[] values, int startIndex, int count);
    //public StringBuilder AppendJoin(char separator, IEnumerable<string> values);
    //public StringBuilder AppendJoin(char separator, string[] values, int startIndex, int count);
    // REJECTED in the first API review round:
    //public StringBuilder AppendJoin<T>(string separator, T[] values, int startIndex, int count);
    //public StringBuilder AppendJoin<T>(string separator, params T[] values);
    //public StringBuilder AppendJoin<T>(char separator, params T[] values);
}
```

Note: Motivation for the API approval update is described in https://github.com/dotnet/corefx/issues/3419#issuecomment-273254581 -- adjusting to follow proven working form of 'string.Join', correctly working with most-used overload calls - see usage example in https://github.com/dotnet/corefx/issues/5552#issuecomment-268395367.

#### Compatibility

Potential source-level issue if someone is already using an extension method with the same name.

@danmosemsft Yes, that fixed it. 

It's not a problem for me, but `build-tests` fails four times with this message:

> ZipArchive\zip_netcoreappTests.cs(24,57): error CS1061: 'ZipArchiveEntry' does not contain a definition for 'ExternalAttributes' and no extension method 'ExternalAttributes' accepting a first argument of type 'ZipArchiveEntry' could be found (are you missing a using directive or a
n assembly reference?) [C:\Users\jmusser\Source\Repos\corefx\src\System.IO.Compression\tests\System.IO.Compression.Tests.csproj]

--- a/src/System.Runtime/ref/System.Runtime.cs
+++ b/src/System.Runtime/ref/System.Runtime.cs
@@ -6930,6 +6930,12 @@ namespace System.Text
         public System.Text.StringBuilder AppendFormat(string format, params object[] args) { throw null; }
         public System.Text.StringBuilder AppendLine() { throw null; }
         public System.Text.StringBuilder AppendLine(string value) { throw null; }
+        public StringBuilder AppendJoin(string separator, params object[] values) { throw null; }
+        public StringBuilder AppendJoin<T>(string separator, System.Collections.Generic.IEnumerable<T> values) { throw null; }
+        public StringBuilder AppendJoin(string separator, params string[] values) { throw null; }
+        public StringBuilder AppendJoin(char separator, params object[] values) { throw null; }
+        public StringBuilder AppendJoin<T>(char separator, System.Collections.Generic.IEnumerable<T> values) { throw null; }
+        public StringBuilder AppendJoin(char separator, params string[] values) { throw null; }
         public System.Text.StringBuilder Clear() { throw null; }
         public void CopyTo(int sourceIndex, char[] destination, int destinationIndex, int count) { }
         public int EnsureCapacity(int capacity) { throw null; }

--- a/src/System.Runtime/src/ApiCompatBaseline.uap.txt
+++ b/src/System.Runtime/src/ApiCompatBaseline.uap.txt
@@ -1,0 +1,6 @@
+Compat issues with assembly System.Runtime:
+MembersMustExist : Member 'System.Text.StringBuilder.AppendJoin(System.Char, System.Object[])' does not exist in the implementation but it does exist in the contract.
+MembersMustExist : Member 'System.Text.StringBuilder.AppendJoin(System.Char, System.String[])' does not exist in the implementation but it does exist in the contract.
+MembersMustExist : Member 'System.Text.StringBuilder.AppendJoin(System.String, System.Object[])' does not exist in the implementation but it does exist in the contract.
+MembersMustExist : Member 'System.Text.StringBuilder.AppendJoin(System.String, System.String[])' does not exist in the implementation but it does exist in the contract.
+Total Issues: 4

--- a/src/System.Runtime/src/ApiCompatBaseline.uapaot.txt
+++ b/src/System.Runtime/src/ApiCompatBaseline.uapaot.txt
@@ -5,12 +5,8 @@ MembersMustExist : Member 'System.String.EndsWith(System.Char)' does not exist i
 MembersMustExist : Member 'System.String.Replace(System.String, System.String, System.Boolean, System.Globalization.CultureInfo)' does not exist in the implementation but it does exist in the contract.
 MembersMustExist : Member 'System.String.Replace(System.String, System.String, System.StringComparison)' does not exist in the implementation but it does exist in the contract.
 MembersMustExist : Member 'System.String.StartsWith(System.Char)' does not exist in the implementation but it does exist in the contract.
-MembersMustExist : Member 'System.Text.Encoding.BodyName.get()' does not exist in the implementation but it does exist in the contract.
-MembersMustExist : Member 'System.Text.Encoding.GetEncodings()' does not exist in the implementation but it does exist in the contract.
-MembersMustExist : Member 'System.Text.Encoding.HeaderName.get()' does not exist in the implementation but it does exist in the contract.
-MembersMustExist : Member 'System.Text.Encoding.IsBrowserDisplay.get()' does not exist in the implementation but it does exist in the contract.
-MembersMustExist : Member 'System.Text.Encoding.IsBrowserSave.get()' does not exist in the implementation but it does exist in the contract.
-MembersMustExist : Member 'System.Text.Encoding.IsMailNewsDisplay.get()' does not exist in the implementation but it does exist in the contract.
-MembersMustExist : Member 'System.Text.Encoding.IsMailNewsSave.get()' does not exist in the implementation but it does exist in the contract.
-MembersMustExist : Member 'System.Text.Encoding.WindowsCodePage.get()' does not exist in the implementation but it does exist in the contract.
-Total Issues: 14
+MembersMustExist : Member 'System.Text.StringBuilder.AppendJoin(System.Char, System.Object[])' does not exist in the implementation but it does exist in the contract.
+MembersMustExist : Member 'System.Text.StringBuilder.AppendJoin(System.Char, System.String[])' does not exist in the implementation but it does exist in the contract.
+MembersMustExist : Member 'System.Text.StringBuilder.AppendJoin(System.String, System.Object[])' does not exist in the implementation but it does exist in the contract.
+MembersMustExist : Member 'System.Text.StringBuilder.AppendJoin(System.String, System.String[])' does not exist in the implementation but it does exist in the contract.
+Total Issues: 10

--- a/src/System.Runtime/tests/System.Runtime.Tests.csproj
+++ b/src/System.Runtime/tests/System.Runtime.Tests.csproj
@@ -137,6 +137,7 @@
     <Compile Include="System\Runtime\CompilerServices\RuntimeHelpersTests.netcoreapp.cs" />
     <Compile Include="System\Runtime\CompilerServices\ConditionalWeakTableTests.netcoreapp.cs" />
     <Compile Include="System\Runtime\ExceptionServices\ExceptionDispatchInfoTests.netcoreapp.cs" />
+    <Compile Include="System\Text\StringBuilderTests.netcoreapp.cs" />    
     <Compile Include="System\UIntPtrTests.netcoreapp.cs" />
     <Compile Include="System\ComponentModel\DefaultValueAttributeTests.netcoreapp.cs" />
     <Compile Include="System\TypedReferenceTests.cs" />

--- a/src/System.Runtime/tests/System/Text/StringBuilderTests.cs
+++ b/src/System.Runtime/tests/System/Text/StringBuilderTests.cs
@@ -4,14 +4,12 @@
 
 using System.Collections.Generic;
 using System.Globalization;
-using System.Linq;
 using System.Tests;
-using System.Text;
 using Xunit;
 
 namespace System.Text.Tests
 {
-    public static class StringBuilderTests
+    public static partial class StringBuilderTests
     {
         private static readonly string s_chunkSplitSource = new string('a', 30);
         private static readonly string s_noCapacityParamName = PlatformDetection.IsFullFramework ? "requiredLength" : "valueCount";
@@ -705,7 +703,7 @@ namespace System.Text.Tests
 
             // Length is positive
             yield return new object[] { "Hello", null, ", Foo {0,2}", new object[] { "Bar" }, "Hello, Foo Bar" }; // MiValue's length > minimum length (so don't prepend whitespace)
-            yield return new object[] { "Hello", null, ", Foo {0,3}", new object[] { "B" }, "Hello, Foo   B" }; // Value's length < minimum length (so prepend whitespace)            
+            yield return new object[] { "Hello", null, ", Foo {0,3}", new object[] { "B" }, "Hello, Foo   B" }; // Value's length < minimum length (so prepend whitespace)
             yield return new object[] { "Hello", null, ", Foo {0,     3}", new object[] { "B" }, "Hello, Foo   B" }; // Same as above, but verify AppendFormat ignores whitespace
             yield return new object[] { "Hello", null, ", Foo {0,0}", new object[] { "Bar" }, "Hello, Foo Bar" }; // Minimum length is 0
 
@@ -928,79 +926,6 @@ namespace System.Text.Tests
             Assert.Throws<ArgumentOutOfRangeException>(s_noCapacityParamName, () => builder.AppendLine());
             Assert.Throws<ArgumentOutOfRangeException>(s_noCapacityParamName, () => builder.AppendLine("a"));
         }
-
-
-        #region AppendJoin
-
-
-        [Fact]
-        public static void AppendJoin_NullValues_ThrowsArgumentNullException()
-        {
-            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin('|', (object[])null));
-            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin('|', (IEnumerable<object>)null));
-            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin('|', (string[])null));
-            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin("|", (object[])null));
-            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin("|", (IEnumerable<object>)null));
-            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin("|", (string[])null));
-        }
-
-        [Theory]
-        [InlineData(new object[0], "")]
-        [InlineData(new object[] { null }, "")]
-        [InlineData(new object[] { 10 }, "10")]
-        [InlineData(new object[] { null, null }, "|")]
-        [InlineData(new object[] { null, 20 }, "|20")]
-        [InlineData(new object[] { 10, null }, "10|")]
-        [InlineData(new object[] { 10, 20 }, "10|20")]
-        [InlineData(new object[] { null, null, null }, "||")]
-        [InlineData(new object[] { null, null, 30 }, "||30")]
-        [InlineData(new object[] { null, 20, null }, "|20|")]
-        [InlineData(new object[] { null, 20, 30 }, "|20|30")]
-        [InlineData(new object[] { 10, null, null }, "10||")]
-        [InlineData(new object[] { 10, null, 30 }, "10||30")]
-        [InlineData(new object[] { 10, 20, null }, "10|20|")]
-        [InlineData(new object[] { 10, 20, 30 }, "10|20|30")]
-        [InlineData(new object[] { "" }, "")]
-        [InlineData(new object[] { "", "" }, "|")]
-        public static void AppendJoin_TestValues(object[] values, string expected)
-        {
-            var stringValues = Array.ConvertAll(values, _ => _?.ToString());
-            var enumerable = values.Select(_ => _);
-
-            Assert.Equal(expected, new StringBuilder().AppendJoin('|', values).ToString());
-            Assert.Equal(expected, new StringBuilder().AppendJoin('|', enumerable).ToString());
-            Assert.Equal(expected, new StringBuilder().AppendJoin('|', stringValues).ToString());
-            Assert.Equal(expected, new StringBuilder().AppendJoin("|", values).ToString());
-            Assert.Equal(expected, new StringBuilder().AppendJoin("|", enumerable).ToString());
-            Assert.Equal(expected, new StringBuilder().AppendJoin("|", stringValues).ToString());
-        }
-
-        [Fact]
-        public static void AppendJoin_NullToStringValues()
-        {
-            AppendJoin_TestValues(new object[] { new NullToStringObject() }, "");
-            AppendJoin_TestValues(new object[] { new NullToStringObject(), new NullToStringObject() }, "|");
-        }
-
-        private sealed class NullToStringObject
-        {
-            public override string ToString() => null;
-        }
-
-        [Theory]
-        [InlineData(null, "123")]
-        [InlineData("", "123")]
-        [InlineData(" ", "1 2 3")]
-        [InlineData(", ", "1, 2, 3")]
-        public static void AppendJoin_TestStringSeparators(string separator, string expected)
-        {
-            Assert.Equal(expected, new StringBuilder().AppendJoin(separator, new object[] { 1, 2, 3 }).ToString());
-            Assert.Equal(expected, new StringBuilder().AppendJoin(separator, Enumerable.Range(1, 3)).ToString());
-            Assert.Equal(expected, new StringBuilder().AppendJoin(separator, new string[] { "1", "2", "3" }).ToString());
-        }
-
-        #endregion
-
 
         [Fact]
         public static void Clear()

--- a/src/System.Runtime/tests/System/Text/StringBuilderTests.cs
+++ b/src/System.Runtime/tests/System/Text/StringBuilderTests.cs
@@ -975,6 +975,18 @@ namespace System.Text.Tests
             Assert.Equal(expected, new StringBuilder().AppendJoin("|", stringValues).ToString());
         }
 
+        [Fact]
+        public static void AppendJoin_NullToStringValues()
+        {
+            AppendJoin_TestValues(new object[] { new NullToStringObject() }, "");
+            AppendJoin_TestValues(new object[] { new NullToStringObject(), new NullToStringObject() }, "|");
+        }
+
+        private sealed class NullToStringObject
+        {
+            public override string ToString() => null;
+        }
+
         [Theory]
         [InlineData(null, "123")]
         [InlineData("", "123")]

--- a/src/System.Runtime/tests/System/Text/StringBuilderTests.cs
+++ b/src/System.Runtime/tests/System/Text/StringBuilderTests.cs
@@ -960,6 +960,8 @@ namespace System.Text.Tests
         [InlineData(new object[] { 10, null, 30 }, "10||30")]
         [InlineData(new object[] { 10, 20, null }, "10|20|")]
         [InlineData(new object[] { 10, 20, 30 }, "10|20|30")]
+        [InlineData(new object[] { "" }, "")]
+        [InlineData(new object[] { "", "" }, "|")]
         public static void AppendJoin_TestValues(object[] values, string expected)
         {
             var stringValues = Array.ConvertAll(values, _ => _?.ToString());

--- a/src/System.Runtime/tests/System/Text/StringBuilderTests.cs
+++ b/src/System.Runtime/tests/System/Text/StringBuilderTests.cs
@@ -4,6 +4,7 @@
 
 using System.Collections.Generic;
 using System.Globalization;
+using System.Linq;
 using System.Tests;
 using System.Text;
 using Xunit;
@@ -927,6 +928,65 @@ namespace System.Text.Tests
             Assert.Throws<ArgumentOutOfRangeException>(s_noCapacityParamName, () => builder.AppendLine());
             Assert.Throws<ArgumentOutOfRangeException>(s_noCapacityParamName, () => builder.AppendLine("a"));
         }
+
+
+        #region AppendJoin
+
+
+        [Fact]
+        public static void AppendJoin_NullValues_ThrowsArgumentNullException()
+        {
+            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin('|', (object[])null));
+            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin('|', (IEnumerable<object>)null));
+            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin('|', (string[])null));
+            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin("|", (object[])null));
+            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin("|", (IEnumerable<object>)null));
+            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin("|", (string[])null));
+        }
+
+        [Theory]
+        [InlineData(new object[0], "")]
+        [InlineData(new object[] { null }, "")]
+        [InlineData(new object[] { 10 }, "10")]
+        [InlineData(new object[] { null, null }, "|")]
+        [InlineData(new object[] { null, 20 }, "|20")]
+        [InlineData(new object[] { 10, null }, "10|")]
+        [InlineData(new object[] { 10, 20 }, "10|20")]
+        [InlineData(new object[] { null, null, null }, "||")]
+        [InlineData(new object[] { null, null, 30 }, "||30")]
+        [InlineData(new object[] { null, 20, null }, "|20|")]
+        [InlineData(new object[] { null, 20, 30 }, "|20|30")]
+        [InlineData(new object[] { 10, null, null }, "10||")]
+        [InlineData(new object[] { 10, null, 30 }, "10||30")]
+        [InlineData(new object[] { 10, 20, null }, "10|20|")]
+        [InlineData(new object[] { 10, 20, 30 }, "10|20|30")]
+        public static void AppendJoin_TestValues(object[] values, string expected)
+        {
+            var stringValues = Array.ConvertAll(values, _ => _?.ToString());
+            var enumerable = values.Select(_ => _);
+
+            Assert.Equal(expected, new StringBuilder().AppendJoin('|', values).ToString());
+            Assert.Equal(expected, new StringBuilder().AppendJoin('|', enumerable).ToString());
+            Assert.Equal(expected, new StringBuilder().AppendJoin('|', stringValues).ToString());
+            Assert.Equal(expected, new StringBuilder().AppendJoin("|", values).ToString());
+            Assert.Equal(expected, new StringBuilder().AppendJoin("|", enumerable).ToString());
+            Assert.Equal(expected, new StringBuilder().AppendJoin("|", stringValues).ToString());
+        }
+
+        [Theory]
+        [InlineData(null, "123")]
+        [InlineData("", "123")]
+        [InlineData(" ", "1 2 3")]
+        [InlineData(", ", "1, 2, 3")]
+        public static void AppendJoin_TestStringSeparators(string separator, string expected)
+        {
+            Assert.Equal(expected, new StringBuilder().AppendJoin(separator, new object[] { 1, 2, 3 }).ToString());
+            Assert.Equal(expected, new StringBuilder().AppendJoin(separator, Enumerable.Range(1, 3)).ToString());
+            Assert.Equal(expected, new StringBuilder().AppendJoin(separator, new string[] { "1", "2", "3" }).ToString());
+        }
+
+        #endregion
+
 
         [Fact]
         public static void Clear()

--- a/src/System.Runtime/tests/System/Text/StringBuilderTests.netcoreapp.cs
+++ b/src/System.Runtime/tests/System/Text/StringBuilderTests.netcoreapp.cs
@@ -1,0 +1,79 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace System.Text.Tests
+{
+    public static partial class StringBuilderTests
+    {
+        [Fact]
+        public static void AppendJoin_NullValues_ThrowsArgumentNullException()
+        {
+            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin('|', (object[])null));
+            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin('|', (IEnumerable<object>)null));
+            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin('|', (string[])null));
+            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin("|", (object[])null));
+            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin("|", (IEnumerable<object>)null));
+            Assert.Throws<ArgumentNullException>("values", () => new StringBuilder().AppendJoin("|", (string[])null));
+        }
+
+        [Theory]
+        [InlineData(new object[0], "")]
+        [InlineData(new object[] { null }, "")]
+        [InlineData(new object[] { 10 }, "10")]
+        [InlineData(new object[] { null, null }, "|")]
+        [InlineData(new object[] { null, 20 }, "|20")]
+        [InlineData(new object[] { 10, null }, "10|")]
+        [InlineData(new object[] { 10, 20 }, "10|20")]
+        [InlineData(new object[] { null, null, null }, "||")]
+        [InlineData(new object[] { null, null, 30 }, "||30")]
+        [InlineData(new object[] { null, 20, null }, "|20|")]
+        [InlineData(new object[] { null, 20, 30 }, "|20|30")]
+        [InlineData(new object[] { 10, null, null }, "10||")]
+        [InlineData(new object[] { 10, null, 30 }, "10||30")]
+        [InlineData(new object[] { 10, 20, null }, "10|20|")]
+        [InlineData(new object[] { 10, 20, 30 }, "10|20|30")]
+        [InlineData(new object[] { "" }, "")]
+        [InlineData(new object[] { "", "" }, "|")]
+        public static void AppendJoin_TestValues(object[] values, string expected)
+        {
+            var stringValues = Array.ConvertAll(values, _ => _?.ToString());
+            var enumerable = values.Select(_ => _);
+
+            Assert.Equal(expected, new StringBuilder().AppendJoin('|', values).ToString());
+            Assert.Equal(expected, new StringBuilder().AppendJoin('|', enumerable).ToString());
+            Assert.Equal(expected, new StringBuilder().AppendJoin('|', stringValues).ToString());
+            Assert.Equal(expected, new StringBuilder().AppendJoin("|", values).ToString());
+            Assert.Equal(expected, new StringBuilder().AppendJoin("|", enumerable).ToString());
+            Assert.Equal(expected, new StringBuilder().AppendJoin("|", stringValues).ToString());
+        }
+
+        [Fact]
+        public static void AppendJoin_NullToStringValues()
+        {
+            AppendJoin_TestValues(new object[] { new NullToStringObject() }, "");
+            AppendJoin_TestValues(new object[] { new NullToStringObject(), new NullToStringObject() }, "|");
+        }
+
+        private sealed class NullToStringObject
+        {
+            public override string ToString() => null;
+        }
+
+        [Theory]
+        [InlineData(null, "123")]
+        [InlineData("", "123")]
+        [InlineData(" ", "1 2 3")]
+        [InlineData(", ", "1, 2, 3")]
+        public static void AppendJoin_TestStringSeparators(string separator, string expected)
+        {
+            Assert.Equal(expected, new StringBuilder().AppendJoin(separator, new object[] { 1, 2, 3 }).ToString());
+            Assert.Equal(expected, new StringBuilder().AppendJoin(separator, Enumerable.Range(1, 3)).ToString());
+            Assert.Equal(expected, new StringBuilder().AppendJoin(separator, new string[] { "1", "2", "3" }).ToString());
+        }
+    }
+}

--- a/src/System.Runtime/tests/System/Text/StringBuilderTests.netcoreapp.cs
+++ b/src/System.Runtime/tests/System/Text/StringBuilderTests.netcoreapp.cs
@@ -75,5 +75,56 @@ namespace System.Text.Tests
             Assert.Equal(expected, new StringBuilder().AppendJoin(separator, Enumerable.Range(1, 3)).ToString());
             Assert.Equal(expected, new StringBuilder().AppendJoin(separator, new string[] { "1", "2", "3" }).ToString());
         }
+
+
+        private static StringBuilder CreateBuilderWithNoSpareCapacity()
+        {
+            return new StringBuilder(0, 5).Append("Hello");
+        }
+
+        [Theory]
+        [InlineData(null, new object[] { null, null })]
+        [InlineData("", new object[] { "", "" })]
+        [InlineData(" ", new object[] { })]
+        [InlineData(", ", new object[] { "" })]
+        public static void AppendJoin_NoValues_NoSpareCapacity_DoesNotThrow(string separator, object[] values)
+        {
+            var stringValues = Array.ConvertAll(values, _ => _?.ToString());
+            var enumerable = values.Select(_ => _);
+
+            if (separator?.Length == 1)
+            {
+                CreateBuilderWithNoSpareCapacity().AppendJoin(separator[0], values);
+                CreateBuilderWithNoSpareCapacity().AppendJoin(separator[0], enumerable);
+                CreateBuilderWithNoSpareCapacity().AppendJoin(separator[0], stringValues);
+            }
+            CreateBuilderWithNoSpareCapacity().AppendJoin(separator, values);
+            CreateBuilderWithNoSpareCapacity().AppendJoin(separator, enumerable);
+            CreateBuilderWithNoSpareCapacity().AppendJoin(separator, stringValues);
+        }
+
+        [Theory]
+        [InlineData(null, new object[] { " " })]
+        [InlineData(" ", new object[] { " " })]
+        [InlineData(" ", new object[] { null, null })]
+        [InlineData(" ", new object[] { "", "" })]
+        public static void AppendJoin_NoSpareCapacity_ThrowsArgumentOutOfRangeException(string separator, object[] values)
+        {
+            var builder = new StringBuilder(0, 5);
+            builder.Append("Hello");
+
+            var stringValues = Array.ConvertAll(values, _ => _?.ToString());
+            var enumerable = values.Select(_ => _);
+
+            if (separator?.Length == 1)
+            {
+                Assert.Throws<ArgumentOutOfRangeException>(s_noCapacityParamName, () => CreateBuilderWithNoSpareCapacity().AppendJoin(separator[0], values));
+                Assert.Throws<ArgumentOutOfRangeException>(s_noCapacityParamName, () => CreateBuilderWithNoSpareCapacity().AppendJoin(separator[0], enumerable));
+                Assert.Throws<ArgumentOutOfRangeException>(s_noCapacityParamName, () => CreateBuilderWithNoSpareCapacity().AppendJoin(separator[0], stringValues));
+            }
+            Assert.Throws<ArgumentOutOfRangeException>(s_noCapacityParamName, () => CreateBuilderWithNoSpareCapacity().AppendJoin(separator, values));
+            Assert.Throws<ArgumentOutOfRangeException>(s_noCapacityParamName, () => CreateBuilderWithNoSpareCapacity().AppendJoin(separator, enumerable));
+            Assert.Throws<ArgumentOutOfRangeException>(s_noCapacityParamName, () => CreateBuilderWithNoSpareCapacity().AppendJoin(separator, stringValues));
+        }
     }
 }

